Management command to update course seat expires

awaisdar001 opened the pull request:
Create a management command that updates the expires date for all course seat products. This will ensure that, if there are other pages causing this issue, seats for older courses cannot be added to the basket/purchased.

[ECOM-3829](https://openedx.atlassian.net/browse/ECOM-3829)
Related PR: https://github.com/edx/edx-platform/pull/11809

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,110 @@
+        Returns a key, value pair of (course_key, enrollment_end) course and enrollment info.
+        """
+        try:
+            url = get_lms_url('api/courseoverview/enrollment_dates/?exclude_empty={exclude_empty}').format(
```
This is not a real endpoint.

awaisdar001 replied:
@clintonb I created a simple API in courseoverview to get the course enrollment end dates.
https://github.com/edx/edx-platform/pull/11809

clintonb replied:
There are numerous issues with this approach. First, your API is neither versioned nor tested. Second, I'm fairly certain we recently got rid of the `course_overviews` API. Finally, you are returning a lot of information in a single API response. Although, you make a lot of calls, this would be much simpler/cleaner if you used the existing Courses API (https://courses.stage.edx.org/api/courses/v1/courses/course-v1:ASUx+AST111+3T2015/).

awaisdar001 replied:
@clintonb We would need to make requests for each course right? The main idea for returning all data was to make a single request and fetch only required data. 

So seems like we need to call Courses API for each course to fetch information. Right?

clintonb replied:
Yes. Aside from throttling issues, which we can fix for admin users, there is nothing wrong with making multiple calls.

awaisdar001 replied:
Ok, seems like i should switch to courses api.

clintonb replied:
Please do. In general, we should look to use our existing APIs, rather than create new ones. We need to consolidate.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,122 @@
+
+    def handle(self, *args, **options):
+        seats_updated = 0
+        seats_to_update = ['honor', 'audit']
```
What about professional education courses?

awaisdar001 replied:
@clintonb Looking at this.

awaisdar001 commented:
@clintonb @adampalay  Please review.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,121 @@
+
+from django.core.management import BaseCommand
+
+from edx_rest_api_client.client import EdxRestApiClient
```
This import ordering is incorrect. Imports should be grouped into the following categores, and each group should be alphabetized:
* Python
* Third-party
* Project-specific

adampalay replied:
Hey Clinton, is there an ecommerce style guide we should be reviewing before submitting PRs?

clintonb replied:
* https://google.github.io/styleguide/pyguide.html
* https://www.python.org/dev/peps/pep-0008/

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,121 @@
+        if not courses_enrollment_info:
+            return
+
+        child_products = Product.objects.filter(structure=Product.CHILD)
```
Limit to courses product class.

clintonb replied:
Iterating over products is incorrect. You will end up making multiple API calls per course. Iterate over `Course.objects.all()` to cap the number of API calls.

awaisdar001 replied:
@clintonb `self._get_courses_information()` method fetches all the courses information and returns key, value structure `courses_enrollment_info` as `course_id` and `enrollment_end`. 

We are not making additional API calls while executing `courses_enrollment_info.get(seat.course.id)`

clintonb replied:
Okay. Nevertheless, we have more than just course products. You must filter by product class or iterate only over the `Course` objects.

awaisdar001 replied:
I am on my way of updating it to iterate over courses. 
`courses = Course.objects.all()` and then only fetching the seats we are required to update using filtering. and updating them using `.update()` of `Queryset`

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,121 @@
+            return
+
+        child_products = Product.objects.filter(structure=Product.CHILD)
+        logger.info("[%d] Child products found for update.", child_products.count())
```
Lowercase "child".

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,121 @@
+        for course_key in self.enrollment_date_not_found:
+            logger.info("%s", course_key)
+
+    def _get_courses_information(self, **querystring):
```
You are always calling this with `page_size=50`. Hardcode that in the call below, and avoid the hassle of the kwarg.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,121 @@
+        course_enrollments = {}
+        try:
+            page = 1
+            response = api.courses().get(**querystring)
```
Remove this duplicate code. Set `next_page` to some non-false value.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,121 @@
+                course_enrollments.update(enrollment_info)
+            return course_enrollments
+
+        except Exception:  # pylint: disable=broad-except
```
Let the exception bubble up. We cannot recover from it.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,121 @@
+            while next_page:
+                page += 1
+                querystring['page'] = page
+                response = api.courses().get(**querystring)
```
You need to account for throttling errors.

awaisdar001 replied:
Can you please elaborate how can I account the throttling errors?

clintonb replied:
If you get HTTP 429, you will need to pause execution for a few seconds before re-requesting the data.

http://www.django-rest-framework.org/api-guide/throttling/

awaisdar001 replied:
Ok, Couple of questions.
1. How can I know if an API view is handling Throttling? I looked at [CourseListView](https://github.com/edx/edx-platform/blob/5df15fcf0339e8378df3afe0daad60aa408121a0/lms/djangoapps/course_api/views.py#L122) but coudln't get a clue.

2. So seems like upon Throttling, the `response = api.courses().get(**querystring)` will raise an `Exception`? and we need to pause the execution for that time?

clintonb replied:
1. See the link I sent.
2. Catch [`HttpClientError`](https://github.com/samgiles/slumber/blob/master/slumber/exceptions.py#L21) and check the `status_code`.

clintonb commented on ecommerce/extensions/catalogue/tests/test_update_course_seat_expire.py:
```diff
@@ -0,0 +1,145 @@
+import httpretty
+import mock
+from pytz import UTC
+
```
Remove.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,121 @@
+        logger.info("Enrollment missing for: [%d] courses", len(self.enrollment_date_not_found))
+
+        for course_key in self.enrollment_date_not_found:
+            logger.info("%s", course_key)
```
No need for string interpolation.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,121 @@
+                product.expires = parser.parse(enrollment_end_date)
+
+                if save_to_db:
+                    logger.info(
```
This is noisy. Do it once per course, not seat.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,121 @@
+                self.enrollment_date_not_found.add(product.course_id)
+                continue
+
+            if not product.expires and seat_mode in seats_to_update:
```
The existence of `product.expires` doesn't matter. We want to sync the latest data.

clintonb commented:
I will re-review once the management command has been updated.

adampalay commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,121 @@
+
+        # If there is some exception fetching the course information, do not proceed.
+        if not courses_enrollment_info:
+            return
```
throw a `CommandError` instead, to be more descriptive?

adampalay commented on ecommerce/extensions/catalogue/tests/test_update_course_seat_expire.py:
```diff
@@ -0,0 +1,145 @@
+        verified_seat = Product.objects.get(id=verified_seat.id)
+        for seat in seats_expected_to_update:
+            fetched_seat = Product.objects.get(id=seat.id)
+            if commit:
```
I'm not a huge fan of if/else statements in tests — there's not a good way to confirm that every line is being hit in a test otherwise. May make sense just to split out into two tests?

awaisdar001 commented:
@clintonb Updated management command.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,125 @@
+from __future__ import unicode_literals
+from dateutil import parser
+import time
+import logging
```
Alphabetize.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,125 @@
+                self.enrollment_date_not_found.add(course.id)
+                continue
+
+            course_seats = course.seat_products.filter(
```
Move this inside the `if` statement below.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,125 @@
+            logger.error("No course enrollment information found.")
+            return
+
+        courses = Course.objects.all()
```
`order_by('id')`

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,125 @@
+                attribute_values__value_text__in=seats_to_update
+            )
+            if save_to_db:
+                logger.info(
```
Move this after the update operation. The message should be: `Updated expiration date for [{course_id}] seats: [{comma_separated_product_id_list}]`.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,125 @@
+                course_seats.update(expires=expires)
+                seats_updated += len(course_seats)
+
+        logger.info("[%d] seats updated", seats_updated)
```
We don't really need to track to this level of detail.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,125 @@
+
+            # Only proceed if course enrollment information is present
+            if not enrollment_end_date:
+                self.enrollment_date_not_found.add(course.id)
```
Log the error here. Don't wait until after we have iterated over every course.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,125 @@
+                seats_updated += len(course_seats)
+
+        logger.info("[%d] seats updated", seats_updated)
+        self.log_enrollment_missing()
```
Remove this. Log the issue earlier.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,125 @@
+        """
+        Retrieve the enrollment information for all the courses.
+
+        Arguments:
```
Remove.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,125 @@
+            querystring['page'] = page
+            try:
+                response = api.courses().get(**querystring)
+            except HttpClientError, exc:
```
`except HttpClientError as exc:`

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,125 @@
+                # before re-requesting the data.
+                status_code = exc.response.status_code
+                logger.warning("HttpClientError with status_code [%d].", status_code)
+                if status_code == 429:
```
This is the only case we care about. If you get other types of errors, you should raise them, not continue.

```python
if exc.response.status_code != 429:
  raise
else:
  logger.warning('API calls are being rate-limited. Waiting for [%d] seconds before retrying...', self.pause_time)
  time.sleep(self.pause_time)
  page -= 1
  logger.info('Retrying.')
  continue
```

awaisdar001 commented:
@clintonb Comments addressed. Please review.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,116 @@
+        # If not data is fetched from LMS API, do not proceed.
+        if not courses_enrollment_info:
+            logger.error("No course enrollment information found.")
+            return
```
```python
msg = 'No course enrollment information found.'
logger.error(msg)
raise CommandError(msg)
```

Raise an error to ensure the correct exit code is returned. Simply returning, with no exit code, implies the command executed successfully.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,116 @@
+
+            # Only proceed if course enrollment information is present
+            if not enrollment_end_date:
+                logger.error("Enrollment missing for course: [%s]", course.id)
```
No need for the colon.

clintonb replied:
Single quotes.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,116 @@
+                course_seats.update(expires=expires)
+                logger.info(  # todo
+                    "Updated expiration date for [%s] seats: [%s]",
+                    unicode(course),
```
`course.id`

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,116 @@
+                )
+                expires = parser.parse(enrollment_end_date)
+                course_seats.update(expires=expires)
+                logger.info(  # todo
```
Remove the `# todo`.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,116 @@
+            querystring['page'] = page
+            try:
+                response = api.courses().get(**querystring)
+            except HttpClientError as exc:
```
This needs to be tested.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,116 @@
+
+        # If not data is fetched from LMS API, do not proceed.
+        if not courses_enrollment_info:
+            logger.error("No course enrollment information found.")
```
Single quotes.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,116 @@
+            return
+
+        courses = Course.objects.all().order_by('id')
+        logger.info("[%d] courses found for update.", courses.count())
```
Single quotes.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,116 @@
+                expires = parser.parse(enrollment_end_date)
+                course_seats.update(expires=expires)
+                logger.info(  # todo
+                    "Updated expiration date for [%s] seats: [%s]",
```
Single quotes.

clintonb commented on ecommerce/extensions/catalogue/tests/test_update_course_seat_expire.py:
```diff
@@ -0,0 +1,181 @@
+        )
+        self.professional_seat = self.course.create_or_update_seat('professional', False, 0, self.partner)
+
+    def tearDown(self):
```
This is not necessary.

clintonb commented on ecommerce/extensions/catalogue/tests/test_update_course_seat_expire.py:
```diff
@@ -0,0 +1,181 @@
+
+
+@ddt.ddt
+class UpdateSeatExpireDateTests(CourseCatalogTestMixin, TransactionTestCase):
```
Why `TransactionTestCase` instead of `TestCase`?

clintonb commented on ecommerce/extensions/catalogue/tests/test_update_course_seat_expire.py:
```diff
@@ -0,0 +1,181 @@
+    def tearDown(self):
+        super(UpdateSeatExpireDateTests, self).tearDown()
+
+    def _mock_lms_api(self, status, body=None):
```
`mock_courses_api`

clintonb commented on ecommerce/extensions/catalogue/tests/test_update_course_seat_expire.py:
```diff
@@ -0,0 +1,181 @@
+        super(UpdateSeatExpireDateTests, self).tearDown()
+
+    def _mock_lms_api(self, status, body=None):
+        """ Mock LMS API with specific status and body. """
```
s/LMS/Course

clintonb commented on ecommerce/extensions/catalogue/tests/test_update_course_seat_expire.py:
```diff
@@ -0,0 +1,181 @@
+
+    def _mock_lms_api(self, status, body=None):
+        """ Mock LMS API with specific status and body. """
+        self.assertTrue(httpretty.is_enabled(), 'httpretty must be enabled to mock Commerce API calls.')
```
s/Commerce/Course

clintonb commented on ecommerce/extensions/catalogue/tests/test_update_course_seat_expire.py:
```diff
@@ -0,0 +1,181 @@
+
+        # Verify that 'verified' seat has not been updated.
+        verified_seat = Product.objects.get(id=self.verified_seat.id)
+        self.assertEqual(verified_seat.expires, self.expire_date)
```
The verified seat should have a different expiration date; otherwise, you don't truly know if the date was untouched.

awaisdar001 commented:
@clintonb I have changed a little in the management command. 
I am using a new variable `max_tries` and keep track to make finite tries when throttling occurs.
I also added another test `test_update_course_with_exception` as you requested. 

I have addressed the comments also, Please have a look.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,120 @@
+from ecommerce.settings import get_lms_url
+
+logger = logging.getLogger(__name__)
+Course = get_model('courses', 'Course')
```
Load the `Course` model like any other. It's not an Oscar model.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,120 @@
+        save_to_db = options.get('commit', False)
+        courses_enrollment_info = self._get_courses_information()
+
+        # If not data is fetched from LMS API, do not proceed.
```
This comment is not necessary.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,120 @@
+        Retrieve the enrollment information for all the courses.
+
+        Returns:
+            key, value pair of (course_key, enrollment_end) course and enrollment info.
```
Dictionary

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,120 @@
+            except HttpClientError as exc:
+                # this is a known limitation; If we get HTTP429, we need to pause execution for a few seconds
+                # before re-requesting the data. raise any other errors
+                status_code = exc.response.status_code
```
No need for this variable.

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,120 @@
+        course_enrollments = {}
+
+        page = 0
+        throttling_tries = 0
```
`throttling_attempts`

clintonb commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,120 @@
+            page += 1
+            querystring['page'] = page
+            try:
+                response = api.courses().get(**querystring)
```
Set `throttling_attempts` to zero after this call.

clintonb commented:
:+1: once comments addressed, and tests pass.

awaisdar001 commented:
@clintonb Comments addressed. 
@adampalay, @aamir-khan  Please review.

aamir-khan commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,118 @@
+    def handle(self, *args, **options):
+        seats_to_update = ['honor', 'audit', 'no-id-professional', 'professional']
+        save_to_db = options.get('commit', False)
+        courses_enrollment_info = self._get_courses_information()
```
This is not the course's enrollment related information. It is confusing can you rename it to just courses_info

awaisdar001 replied:
Good point Aamir. Actually `self._get_courses_information()` is returning enrollment information, so I think I should name it to something `self._get_courses_enrollment_info()`

aamir-khan commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,118 @@
+            raise CommandError(msg)
+
+        courses = Course.objects.all().order_by('id')
+        logger.info('[%d] courses found for update.', courses.count())
```
I think you should use `len(courses)` instead of count() because count makes a db call.

clintonb replied:
Both make a database call. If you only need the count, use count().

aamir-khan commented on ecommerce/extensions/catalogue/tests/test_update_course_seat_expire.py:
```diff
@@ -0,0 +1,217 @@
+import httpretty
+from slumber.exceptions import HttpClientError
+from testfixtures import LogCapture
+from ecommerce.courses.models import Product
```
Align this to below group

awaisdar001 commented:
@aamir-khan Comments addressed

awaisdar001 commented:
@adampalay @aamir-khan PR updated.

awaisdar001 commented:
@mushtaqak Aamir is out, Can you please take a look.

mushtaqak commented on ecommerce/extensions/catalogue/management/commands/update_course_seat_expire.py:
```diff
@@ -0,0 +1,119 @@
+                    attributes__name='certificate_type',
+                    attribute_values__value_text__in=seats_to_update
+                )
+                expires = parser.parse(enrollment_end_date)
```
I think `if save_to_db:` should be below this line as this is the only thing that is updating the course seat. If update, we should log, seat updated. 

If I do not give commit option, that means I am just dry-running the command. So, I would still want to get the the courses which **could be affected** by this command. The log should always be printed.

mushtaqak commented:
One suggestion, it is upto you if you want to do it :) 

Nice work :+1:  from me too
